storage/fanout.go

```diff
@@ -231,23 +232,66 @@ func NewMergeQuerier(primaryQuerier Querier, queriers []Querier) Querier {
 func (q *mergeQuerier) Select(params *SelectParams, matchers ...*labels.Matcher) (SeriesSet, Warnings, error) {
 	seriesSets := make([]SeriesSet, 0, len(q.queriers))
 	var warnings Warnings
-	for _, querier := range q.queriers {
-		set, wrn, err := querier.Select(params, matchers...)
-		q.setQuerierMap[set] = querier
-		if wrn != nil {
-			warnings = append(warnings, wrn...)
-		}
-		if err != nil {
-			q.failedQueriers[querier] = struct{}{}
-			// If the error source isn't the primary querier, return the error as a warning and continue.
-			if querier != q.primaryQuerier {
-				warnings = append(warnings, err)
-				continue
-			} else {
-				return nil, nil, err
+	if len(q.queriers) > 1 {
+		wg := sync.WaitGroup{}
+		wg.Add(len(q.queriers))
+		queryResult := make(map[int]SeriesSet)
+		var priErr error = nil
+		var lock sync.Mutex
+		var wgLock sync.Mutex
+		for idx, qr := range q.queriers {
+			go func(index int, querier Querier) {
```

Comment:
You could simplify and reduce duplication if you did all the processing of the result outside the goroutine.

Reply:
![image](https://user-images.githubusercontent.com/9583245/73452508-002f8300-43a5-11ea-933c-6d835888325f.png)
Thank you for your patient guidance. I removed some duplicate code according to your suggestion